presto-native-execution/presto_cpp/main/SystemConnector.h

```diff
@@ -0,0 +1,233 @@
+
+  std::unordered_map<std::string, velox::RuntimeCounter> runtimeStats()
+      override {
+    // TODO: Which stats do we want to expose here?
+    return {};
+  }
+
+ private:
+  enum class TaskColumnEnum {
+    // Note: These values are in the same order as SystemTableHandle schema.
+    kNodeId = 0,
+    kTaskId,
+    kStageExecutionId,
+    kStageId,
+    kQueryId,
+    kState,
+    kSplits,
+    kQueuedSplits,
+    kRunningSplits,
+    kCompletedSplits,
+    kSplitScheduledTimeMs,
+    kSplitCpuTimeMs,
+    kSplitBlockedTimeMs,
+    kRawInputBytes,
+    kRawInputRows,
+    kProcessedInputBytes,
+    kProcessedInputRows,
+    kOutputBytes,
+    kOutputRows,
+    kPhysicalWrittenBytes,
+    kCreated,
+    kStart,
+    kLastHeartBeat,
+    kEnd,
+  };
+
+  velox::RowVectorPtr getTaskResults();
+
+  // Mapping between output columns and their indices (column_index_t)
+  // corresponding to the taskInfo fields for them.
+  std::vector<velox::column_index_t> outputColumnMappings_;
+  velox::RowTypePtr outputType_;
+
+  const TaskManager* taskManager_;
+  velox::memory::MemoryPool* pool_;
+
+  std::shared_ptr<SystemSplit> currentSplit_;
+
+  size_t completedRows_{0};
+  size_t completedBytes_{0};
+};
+
+class SystemConnector : public velox::connector::Connector {
+ public:
+  SystemConnector(
+      const std::string& id,
+      std::shared_ptr<const velox::Config> /*properties*/,
+      folly::Executor* /*executor*/)
+      : Connector(id) {}
+
+  void setTaskManager(const TaskManager* taskManager) {
+    taskManager_ = taskManager;
+  }
+
+  std::unique_ptr<velox::connector::DataSource> createDataSource(
+      const velox::RowTypePtr& outputType,
+      const std::shared_ptr<velox::connector::ConnectorTableHandle>&
+          tableHandle,
+      const std::unordered_map<
+          std::string,
+          std::shared_ptr<velox::connector::ColumnHandle>>& columnHandles,
+      velox::connector::ConnectorQueryCtx* connectorQueryCtx) override final {
+    VELOX_CHECK(taskManager_);
+    return std::make_unique<SystemDataSource>(
+        outputType,
+        tableHandle,
+        columnHandles,
+        taskManager_,
+        connectorQueryCtx->memoryPool());
+  }
+
+  std::unique_ptr<velox::connector::DataSink> createDataSink(
+      velox::RowTypePtr /*inputType*/,
+      std::shared_ptr<
+          velox::connector::
+              ConnectorInsertTableHandle> /*connectorInsertTableHandle*/,
+      velox::connector::ConnectorQueryCtx* /*connectorQueryCtx*/,
+      velox::connector::CommitStrategy /*commitStrategy*/) override final {
+    VELOX_NYI("SystemConnector does not support data sink.");
+  }
+
+ private:
+  const TaskManager* taskManager_;
+};
+
+class SystemConnectorFactory : public velox::connector::ConnectorFactory {
+ public:
+  static constexpr const char* kSystemConnectorName{"$system"};
+
+  SystemConnectorFactory() : SystemConnectorFactory(kSystemConnectorName) {}
+
+  explicit SystemConnectorFactory(const char* connectorName)
```

Comment:
Otherwise maybe we can pass the `TaskManager` to the `SystemConnectorFactory` and pass it to the `SystemConnector`?

Reply:
Not needed anymore.